src/main/java/io/harness/cf/client/api/InnerClient.java

```diff
@@ -0,0 +1,275 @@
+    String cluster = (String) untrusted.getBody().get("clusterIdentifier");
+
+    // set values to processors
+    pollProcessor.setEnvironment(environment);
+    pollProcessor.setCluster(cluster);
+
+    streamProcessor.setEnvironment(environment);
+    streamProcessor.setCluster(cluster);
+    streamProcessor.setToken(token);
+
+    metricsProcessor.setEnvironmentID(environment);
+    metricsProcessor.setCluster(cluster);
+    metricsProcessor.setToken(token);
+  }
+
+  protected void onUnauthorized() {
+    authService.startAsync();
+    pollProcessor.stop();
+    if (options.isStreamEnabled()) {
+      streamProcessor.stop();
+    }
+
+    if (options.isAnalyticsEnabled()) {
+      metricsProcessor.stop();
+    }
+  }
+
+  @Override
+  public void onAuthSuccess(@NonNull final String token) {
+    log.info("SDK successfully logged in");
+    processToken(token);
+    // run services only after token is processed
+    pollProcessor.start();
+
+    if (options.isStreamEnabled()) {
+      streamProcessor.start();
+    }
+
+    if (options.isAnalyticsEnabled()) {
+      metricsProcessor.start();
+    }
+  }
+
+  @Override
+  public void onAuthError(String error) {
+    failure = true;
+  }
+
+  @Override
+  public void onPollerReady() {
+    initialize(Processor.POLL);
+  }
+
+  @Override
+  public void onPollerError(@NonNull String error) {
+    failure = true;
+  }
+
+  @Override
+  public void onStreamConnected() {
+    pollProcessor.stop();
+  }
+
+  @Override
+  public void onStreamDisconnected() {
+    pollProcessor.start();
+  }
+
+  @Override
+  public void onStreamReady() {
+    initialize(Processor.STREAM);
+  }
+
+  @Override
+  public void onStreamError() {}
+
+  @Override
+  public void onFlagStored(@NonNull String identifier) {}
+
+  @Override
+  public void onFlagDeleted(@NonNull String identifier) {}
+
+  @Override
+  public void onSegmentStored(@NonNull String identifier) {}
+
+  @Override
+  public void onSegmentDeleted(@NonNull String identifier) {}
+
+  @Override
+  public void onMetricsReady() {
+    initialize(Processor.METRICS);
+  }
+
+  @Override
+  public void onMetricsError(@NonNull String error) {}
+
+  @Override
+  public void onMetricsFailure() {
+    failure = true;
+  }
+
+  private synchronized void initialize(Processor processor) {
+    switch (processor) {
+      case POLL:
+        pollerReady = true;
+        log.debug("PollingProcessor ready");
+        break;
+      case STREAM:
+        streamReady = true;
+        log.debug("StreamingProcessor ready");
+        break;
+      case METRICS:
+        metricReady = true;
+        log.debug("MetricsProcessor ready");
+        break;
+    }
+
+    if (options.isStreamEnabled() && !streamReady) {
+      return;
+    }
+
+    if (options.isAnalyticsEnabled() && !this.metricReady) {
+      return;
+    }
+
+    if (!this.pollerReady) {
+      return;
+    }
+
+    initialized = true;
+    notify();
+    log.info("Initialization is complete");
```

Comment:
This might be wrong. Would the initialisation be complete after any processor is ready or all of them? Also this method looks like it should be broken into 3 methods, one per processor.

Reply:
initialisation is complete when all processor are ready